### PR TITLE
chore: Add Audit Log components and service to load from the API

### LINKED ISSUE #3607: Display audit log diffs in a git diff format
As a user, I would like to see audit log diffs so I understand what values have changed after certain events.

[Audit RFC](https://www.notion.so/coderhq/v2-Enterprise-Audit-Logging-a6358d50a453413eae6441e689f023ac)

[Audit Mocks](https://www.figma.com/file/oy7echtkBVRuOCMBVSrr4T/Auditing?node-id=0%3A1)

## Technical Considerations
This work is dependent upon the log scaffolded out in #3429. The expandable section detailed in that ticket will need to be created before it can be populated.

We can use [react-diff-viewer](https://github.com/praneshr/react-diff-viewer) to create the below audit diff visual:

Screen Shot 2022-08-19 at 3 55 16 PM

Only certain events should have a diff section - see below AC

## AC

- [ ] Populate the expandable section with diffs returned from `/api/v2/audit`
- [ ] If a log event has no diffs (ex. a user logged in), there will be no expandable section. The top-level human-readable string would read something like “Alice Harvey logged in”. The human readable string will be returned from the back end.
- [ ] If a log event has 1 diff (ex. a user edited the TTL on a workspace), there will be no expandable section. The top-level human-readable string would read something like ”Alice Harvey updated the TTL on ‘workspace-x’ from x to y.”
- [ ] If a log event has multiple diffs (ex. a user edited multiple fields on a template), there will be an expandable section which can be toggled to reveal the diffs. The top level human-readable string would contain the first diff and read something like ”Alice Harvey renamed ‘template-a’ to ‘template-b”. The expandable diff section would contain additional diffs, e.g.
    
    ```jsx
    [01] Updated name from template-a to template-b
    [02] Updated description from ‘foo’ to ‘bar’
    ```

- [ ] Add tests and stories

## Definition of Done
When I get diffs from the API,
then they are represented in a git diff in an expandable section.

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Preview:
Screen Shot 2022-08-31 at 15 32 31

Demo:

https://user-images.githubusercontent.com/3165839/188952526-5215de23-fbb6-4847-8431-8b53c26f6846.mov

⚠️ Technical considerations:
- I decided to build a diff component instead of getting Monaco or any other editor. Having a custom component was easier to customize and easier to get the format - which is not JSON or any other language but a list of values - done.
- Some resources don't have an internal link like User or TemplateVersion so instead of being a link/blue, they are strong text.
- I used test-ids to test the page loading since the message is pretty dynamic
- I used the HTTP status number since it can be helpful to debug. The colors will change accordingly.
- As "extra data" on the right I just put the IP and the Agent. I'm not sure if we should add other info.
- The collapse button only works if the log has diffs and it works with keyboard navigation as well.

Closes #3607 #3429 #3428 #3608

### CODE REVIEW

File: site/src/api/typesGenerated.ts [hunk 1]
```diff
@@ -620,3 +620,27 @@ export type WorkspaceAgentStatus = "connected" | "connecting" | "disconnected"
 
 // From codersdk/workspacebuilds.go
 export type WorkspaceTransition = "delete" | "start" | "stop"
+
+// TODO: Remove this when the generated types work for AuditLogs
+export interface AuditLog {
```
Comment (presleyp) [inline site/src/api/typesGenerated.ts:625]: Will this get overwritten when types are regenerated? Maybe it should go in types.ts instead?

Comment (BrunoQuaresma) [reply site/src/api/typesGenerated.ts:625]: It should be overwritten!

File: site/src/components/AuditLogRow/AuditLogDiff.tsx [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+const getDiffValue = (value: number | string | boolean) => {
+  if (typeof value === "string") {
+    return `"${value}"`
+  }
+
+  return value.toString()
+}
+
+export const AuditDiff: React.FC<{ diff: AuditLog["diff"] }> = ({ diff }) => {
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogDiff.tsx:14]: Looks like the file is AuditLogDiff but the component is AuditDiff, could you make them the same?

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+import { Link as RouterLink } from "react-router-dom"
+import { createDayString } from "util/createDayString"
+import { AuditDiff } from "./AuditLogDiff"
+
+const pillTypeByHttpStatus = (httpStatus: number): ComponentProps<typeof Pill>["type"] => {
+  if (httpStatus >= 300 && httpStatus < 500) {
+    return "warning"
+  }
+
+  if (httpStatus > 500) {
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:21]: I think this should be >=

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    return resource.name
+  }
+
+  return resource.username
+}
+
+const getResourceHref = (
+  resource: AuditLog["resource"],
+  resourceType: AuditLog["resource_type"],
+): string | undefined => {
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:39]: Really clean!

File: site/src/components/AuditLogRow/AuditLogDiff.tsx (hunk 2 above)
Comment (BrunoQuaresma) [reply site/src/components/AuditLogRow/AuditLogDiff.tsx:14]: Sure! I missed this.

File: site/src/components/AuditLogRow/AuditLogRow.tsx (hunk 3 above)
Comment (BrunoQuaresma) [reply site/src/components/AuditLogRow/AuditLogRow.tsx:21]: Yes!

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+      {label}
+    </Link>
+  )
+}
+
+const actionLabelByAction: Record<AuditLog["action"], string> = {
+  create: "created a new",
+  write: "updated",
+  delete: "deleted",
+}
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:74]: Can you use i18n for this?

Comment (BrunoQuaresma) [reply site/src/components/AuditLogRow/AuditLogRow.tsx:74]: I think so!

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                  <strong>IP</strong> {auditLog.ip}
+                </div>
+                <div>
+                  <strong>Agent</strong> {auditLog.user_agent}
+                </div>
+              </Stack>
+            </Stack>
+          </Stack>
+
+          <div className={shouldDisplayDiff ? undefined : styles.disabledDropdownIcon}>
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:167]: I wonder if it's better to omit the dropdown icon when there's no diff? Also just fyi, I made an Expander component recently that labels the arrows "More" and "Less". It might be too wordy for the audit log though!

File: site/src/xServices/audit/auditXService.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    },
+    tsTypes: {} as import("./auditXService.typegen").Typegen0,
+    initial: "loadingLogs",
+    states: {
+      loadingLogs: {
+        invoke: {
+          src: "loadAuditLogs",
+          onDone: {
+            target: "success",
+            actions: ["assignAuditLogs"],
```
Comment (presleyp) [inline site/src/xServices/audit/auditXService.ts:25]: What happens in case of error?

File: site/src/pages/TemplatePage/TemplatePageView.tsx [hunk 8]
```diff
@@ -56,7 +56,7 @@ export const TemplatePageView: FC<React.PropsWithChildren<TemplatePageViewProps>
     <Margins>
       <PageHeader
         actions={
-          <Stack direction="row" spacing={1}>
+          <>
```
Comment (presleyp) [inline site/src/pages/TemplatePage/TemplatePageView.tsx:59]: Why this change?

Comment (BrunoQuaresma) [reply site/src/pages/TemplatePage/TemplatePageView.tsx:59]: The PageHeader is already wrapping the actions inside of a Stack.

File: site/src/xServices/audit/auditXService.ts (hunk 7 above)
Comment (BrunoQuaresma) [reply site/src/xServices/audit/auditXService.ts:25]: I forgot to handle the error 🤦 going to fix

**Comment by presleyp** [comment]
Love your diff component! Small issue though, the line numbers aren't aligned - I guess it has to do with the text wrapping? 
Screen Shot 2022-08-31 at 3 19 08 PM

File: site/src/components/AuditLogRow/AuditLogRow.tsx (hunk 6 above)
Comment (BrunoQuaresma) [reply site/src/components/AuditLogRow/AuditLogRow.tsx:167]: > I wonder if it's better to omit the dropdown icon when there's no diff? 

I like to keep it and just disable it to avoid "layout shift"/different layout from a component that is expandable and another that is not, but no strong opinion.

> Also just fyi, I made an Expander component recently that labels the arrows "More" and "Less". It might be too wordy for the audit log though!

Nice! I think they handle diff cases. Here, the row should be clickable. The icon is more of a "hint" for the user that they can click or interact with the element.

**Comment by BrunoQuaresma** [comment]
> Love your diff component! Small issue though, the line numbers aren't aligned - I guess it has to do with the text wrapping? Screen Shot 2022-08-31 at 3 19 08 PM

Good catch! I will play around with smaller viewports and see what other bugs I can fix. Thanks.

**Comment by presleyp** [comment]
I like the color-coded pills for status codes, but wonder if it will be unclear to some users what they represent. Maybe they could say `Status 200` or `Status: 200` instead of just `200`.

**Comment by presleyp** [comment]
I denied some of the Chromatic changes - it's all to do with indentations making values not line up with their headers. Not sure what caused it.

**Comment by kylecarbs** [comment]
Does this handle pagination?

**Comment by BrunoQuaresma** [comment]
> Does this handle pagination?

Nops, but I can do it tomorrow - it would be a diff PR tho. What do you think? Or the port thing should be a priority?

**Comment by BrunoQuaresma** [comment]
> I like the color-coded pills for status codes, but wonder if it will be unclear to some users what they represent. Maybe they could say `Status 200` or `Status: 200` instead of just `200`.

I think HTTP status is very well known 🤔 . For this layout specifically, adding the Status on every row can be too much visually because it increases the "accent color" area. I would wait for users to complain about that. But, if you think we should, I would be happy to add. Here is a comparison:

Screen Shot 2022-08-31 at 20 38 50

Screen Shot 2022-08-31 at 20 39 23

**Comment by presleyp** [comment]
Sure, let's leave the status code as just the number and see if people ask about it.

**Comment by presleyp** [comment]
I denied a few more Chromatic changes, this time it's just that a color change you made affected the Schedule switch, I don't think it looks "on" enough in the dark blue-grey color.

File: site/src/xServices/audit/auditXService.ts (hunk 7 above)
Comment (BrunoQuaresma) [reply site/src/xServices/audit/auditXService.ts]: Added.

File: site/src/xServices/audit/auditXService.ts [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+      }),
+      assignPage: assign({
+        page: ({ page }) => page,
+      }),
+      displayLoadAuditLogsError: (_, event) => {
+        const message = getErrorMessage(event.data, "Error on loading audit logs.")
+        displayError(message)
+      },
+      displayLoadAuditLogsCountError: (_, event) => {
+        const message = getErrorMessage(event.data, "Error on loading audit logs count.")
```
Comment (presleyp) [inline site/src/xServices/audit/auditXService.ts:104]:
```suggestion
        const message = getErrorMessage(event.data, "Error on loading number of audit log entries.")
```
I thought this was a little clearer!

**Review by presleyp — APPROVED** [review]

**Comment by BrunoQuaresma** [comment]
I made some minor updates:
- on auditXService I updated the invoke to use only one function that loads the count and the results. Having multiple invokes doesn't guarantee both services are completed at the end so we would have to use parallel states making it more complex than necessary only to load the count.
- The URL also changes on pagination so it can be a shareable URL

File: site/src/components/AuditLogRow/AuditLogRow.tsx [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+  template: "template",
+  template_version: "template version",
+  user: "user",
+  workspace: "workspace",
+}
+
+const readableActionMessage = (auditLog: AuditLog) => {
+  return `${actionLabelByAction[auditLog.action]} ${
+    resourceLabelByResourceType[auditLog.resource_type]
+  }`
```
Comment (presleyp) [inline site/src/components/AuditLogRow/AuditLogRow.tsx:46]: Here's how to do interpolation with react-i18n https://www.i18next.com/translation-function/interpolation

Comment (BrunoQuaresma) [reply site/src/components/AuditLogRow/AuditLogRow.tsx:46]: We are planning to move this "message builder" to the BE so the CLI and FE can use the same language so I'm going just let this as it is but thanks for sharing the docs. Do you know how they could support interpolation with components?

Comment (presleyp) [reply site/src/components/AuditLogRow/AuditLogRow.tsx:46]: This is the closest I've found https://www.i18next.com/translation-function/formatting
